source/class/qx/Bootstrap.js

```diff
@@ -15,723 +16,505 @@
-    parent[part] = object;
+// To be imlemented in the future
+let FUTURE = false;
 
-    // return last part name (e.g. classname)
-    return part;
-  },
+// Bootstrap the Bootstrap static class
+window.qx = Object.assign(window.qx || {}, {
+  $$namespaceRoot: window,
 
-  setDisplayName(fcn, classname, name) {
-    fcn.displayName = classname + "." + name + "()";
-  },
+  Bootstrap: {
+    /** Undeclared member variables we've already notified the user of */
+    $$undeclared: {},
 
-  setDisplayNames(functionMap, classname) {
-    for (var name in functionMap) {
-      var value = functionMap[name];
-      if (value instanceof Function) {
-        value.displayName = classname + "." + (value.name || name) + "()";
-      }
-    }
-  },
+    /** @type {Map} Stores all defined classes */
+    $$registry: {},
+
+    _allowedStaticKeys: {
+      "@": "object",
+      type: "string", // String
+      include: "object", // Mixin[]
+      statics: "object", // Map
+      environment: "object", // Map
+      events: "object", // Map
+      defer: "function" // Function
+    },
 
-  base(args, varargs) {
-    if (qx.Bootstrap.DEBUG) {
-      if (!qx.Bootstrap.isFunction(args.callee.base)) {
+    _allowedNonStaticKeys: {
+      "@": "object",
+      "@construct": "object",
+      "@destruct": "object",
+      type: "string", // String
+      extend: "function", // Function
+      // extendNativeClass: "boolean", // Boolean -- Does not work yet
+      implement: "object", // Interface[]
+      include: "object", // Mixin[]
+      construct: "function", // Function
+      statics: "object", // Map
+      properties: "object", // Map
+      members: "object", // Map
+      environment: "object", // Map
+      events: "object", // Map
+      delegate: "object", // Map
+      defer: "function", // Function
+      destruct: "function" // Function
+    },
+
+    define(className, config) {
+      let clazz;
+      let proxy;
+      let handler;
+      let path;
+      let classnameComponents;
+      let implicitType = false;
+
+      // Ensure the desginated class has not already been defined
+      if (className && qx.Bootstrap.$$registry[className]) {
         throw new Error(
-          "Cannot call super class. Method is not derived: " +
-            args.callee.displayName
+          `${className} is already defined; cannot redefine a class`
         );
       }
-    }
 
-    if (arguments.length === 1) {
-      return args.callee.base.call(this);
-    } else {
-      return args.callee.base.apply(
-        this,
-        Array.prototype.slice.call(arguments, 1)
-      );
-    }
-  },
-
-  define(name, config) {
-    var isStrictMode = function () {
-      return typeof this == "undefined";
-    };
-
-    if (!config) {
-      config = { statics: {} };
-    }
+      // Process environment
+      let environment = config.environment || {};
+      for (let key in environment) {
+        qx["core"]["Environment"].add(key, environment[key]);
```

Comment:
```suggestion
        qx.core.Environment.add(key, environment[key]);
```

Reply:
We can't do that in qx.Bootstrap because it would bring in qx.core.Environment before it's ready. I originally intended to merge Bootstrap and Class, and thus Class had a bunch of these too. I did correct those recently, but the ones in Bootstrap have to remain as is.